pkg/kubernetes/rollout/v1alpha1/rolloutstatus.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2019 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	api_apps_v1 "k8s.io/api/apps/v1"
+	api_extn_v1beta1 "k8s.io/api/extensions/v1beta1"
+)
+
+// rolloutOutput struct contaons message and boolean value to show rolloutstatus
+type rolloutOutput struct {
+	IsRolledout bool   `json:"IsRolledout"`
+	Message     string `json:"Message"`
+}
+
+// ExtnV1Beta1DeployRolloutStatusGenerator generates RolloutOutput from deploymnet object for that deployment
+var ExtnV1Beta1DeployRolloutStatusGenerator = func(deploy api_extn_v1beta1.Deployment) (op []byte, err error) {
+	r := rolloutOutput{}
+	r.IsRolledout = false
+	var cond *api_extn_v1beta1.DeploymentCondition
+	// list all conditions and and select that condition which type is Progressing.
+	for i := range deploy.Status.Conditions {
+		c := deploy.Status.Conditions[i]
+		if c.Type == api_extn_v1beta1.DeploymentProgressing {
+			cond = &c
+		}
+	}
+	// if deploy.Generation <= deploy.Status.ObservedGeneration then deployment spec is not updated yet.
+	// it marked IsRolledout as false and update message accordingly
+	if deploy.Generation <= deploy.Status.ObservedGeneration {
+		// If Progressing condition's reason is ProgressDeadlineExceeded then it is not rolled out.
+		if cond != nil && cond.Reason == "ProgressDeadlineExceeded" {
+			r.Message = fmt.Sprintf("Deployment exceeded its progress deadline")
```

Comment:
It seems that message is very crucial. So we might want below logic:
```go
var rolloutMessages := map[predicate]string{
  IsProgressDeadlineExceeded: "Deployment exceeded its progress deadline",
  IsRolloutInProgress: "waiting for rollout to finish: %d of %d new replicas updated",
  IsTerminationInProgress: "waiting for termination to finish: %d of %d old replicas terminated",
}

type rolloutMessage(p predicate) string {
  return rolloutMessages[p]
}
```

Reply:
Added this in new package. resolving this conversation. will continue conversation in that package.